sdk/rust/oak/src/handle.rs

```diff
@@ -0,0 +1,224 @@
+/// #   fn visit<F: FnMut(&mut Handle)>(&mut self, mut visitor: F) -> F {
+/// #     visitor(&mut self.handle);
+/// #     visitor
+/// #   }
+/// # }
+///
+/// let mut thing = Thing { handle: 42 };
+///
+/// let handles = extract_handles(&mut thing);
+///
+/// assert_eq!(handles, vec![42]);
+/// ```
+pub fn extract_handles<T: HandleVisit>(msg: &mut T) -> Vec<Handle> {
+    let mut handles = Vec::new();
+    msg.visit(|handle: &mut Handle| {
+        handles.push(*handle);
+        *handle = oak_abi::INVALID_HANDLE;
+    });
+    handles
+}
+
+/// Inject handles into a message.
+///
+/// If the number of handles provided is not exactly equal to the number of handles needed to fill
+/// the message, an error is returned.
+///
+/// ```
+/// use oak::handle::{Handle, HandleVisit, inject_handles};
+///
+/// # #[derive(Debug, PartialEq)]
+/// struct Thing {
+///     handle: Handle,
+/// }
+/// # impl HandleVisit for Thing {
+/// #   fn visit<F: FnMut(&mut Handle)>(&mut self, mut visitor: F) -> F {
+/// #     visitor(&mut self.handle);
+/// #     visitor
+/// #   }
+/// # }
+///
+/// let handles = vec![42];
+/// let mut thing = Thing { handle: 0 };
+///
+/// inject_handles(&mut thing, &handles).unwrap();
+///
+/// assert_eq!(thing, Thing { handle: 42 });
+pub fn inject_handles<T: HandleVisit>(msg: &mut T, handles: &[Handle]) -> Result<(), OakError> {
+    let mut handles = handles.iter();
+    let mut result = Ok(());
+    msg.visit(|handle| {
+        if let Some(to_inject) = handles.next() {
+            *handle = *to_inject;
+        } else {
+            result = Err(OakError::ProtobufDecodeError(None));
+        }
+    });
+    if handles.next().is_some() {
+        result = Err(OakError::ProtobufDecodeError(None));
+    }
+    result
+}
+
+// Import the procedural macro that automatically derives implementations of the trait.
+pub use oak_derive::HandleVisit;
+
+// Implementations for the types generated from different field modifiers
+// (https://github.com/danburkert/prost#scalar-values).
+
+// Optional fields
+impl<T: HandleVisit> HandleVisit for Option<T> {
+    fn visit<F: FnMut(&mut Handle)>(&mut self, visitor: F) -> F {
+        if let Some(inner) = self {
+            inner.visit(visitor)
+        } else {
+            visitor
+        }
+    }
+}
+
+// For repeated fields.
+impl<T: HandleVisit> HandleVisit for Vec<T> {
+    fn visit<F: FnMut(&mut Handle)>(&mut self, visitor: F) -> F {
+        self.iter_mut()
+            .fold(visitor, |visitor, item| item.visit(visitor))
+    }
+}
+
+// For recursive messages.
+impl<T: HandleVisit> HandleVisit for Box<T> {
+    fn visit<F: FnMut(&mut Handle)>(&mut self, visitor: F) -> F {
+        self.as_mut().visit(visitor)
+    }
+}
+
+// For maps. This is only supported for maps that have a key implementing `Ord`, because we need to
+// be able to define an order in which to inject/extract handles. Since protobuf only supports
+// integral and string types for keys, having this constraint is fine.
```

Comment:
Is there a source for this? If so, could you add it here? I actually thought you could use arbitrary types (including other messages) as keys.

Reply:
It is mentioned here: https://developers.google.com/protocol-buffers/docs/proto3#maps.

I'll leave that link in a comment section as well.